File: Data Structures & Algorithms/balanced-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        boolean[] balanced = new boolean[1];
        balanced[0] = true;
        dfs(root, balanced);
        return balanced[0];
    }

    private int dfs(TreeNode root, boolean[] balanced) {
        if (root == null) {
            return 0;
        }

        int left = dfs(root.left, balanced);
        int right = dfs(root.right, balanced);
        if (left - right > 1 || right - left > 1) {
            balanced[0] = false;
        }
        return 1 + Math.max(left, right);
    }
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        
        if not root:
            return None
        
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            else:
                rep = root.right
                while rep.left:
                    rep = rep.left
                
                tmpRight = root.right
                tmpLeft = root.left
                root = TreeNode(rep.val)
                root.left = tmpLeft
                root.right = tmpRight

                root.right = self.deleteNode(root.right, rep.val)
        
        return root
